driver.go

```diff
@@ -81,7 +81,8 @@ func (d MySQLDriver) Open(dsn string) (driver.Conn, error) {
 		}
 	}
 
-	mc.buf = newBuffer(mc.netConn)
+	// Make the buffer get the context from mysqlConn.
+	mc.buf = newBufferContext(mc.netConn, &ContextWrapper{mc: mc})
```

Comment:
buffer.fill() sets the timeout by SetReadDeadline() and mysqlConn.writePacket() sets the timeout by SetWriteDeadline. Context.Deadline() may be earlier than those two timeouts. So Context.Deadline() must be passed into these two functions. There are two ways to pass it. The first one is to pass it as a parameter of a function. The second one is to pass it as a member of a struct. If I chose the first one, I would have to modify every functions from mysqlConn.Ping() to buffer.fill() and mysqlConn.writePacket(). The second one is much easier and there is no need to wrap Context by something like context.withCancel() in this case.

The real context is stored at mysqlConn.ctx and the buffer.ctx is just a wrapper/handler which points to the previous one. I use wrapper/handler because copying a value everywhere is not a good idea. People may modify some and forget the others.